src/Credentials/AssumeRoleCredentialResolver.php

```diff
@@ -0,0 +1,178 @@
+            return self::reject("Profile specified is not an assume role profile.");
+        }
+
+        $sourceCreds = $this->resolveSourceProfile();
+        $region = $this->resolveRegion();
+        if (empty($region)) {
+            return self::reject("'region' must be provided to retrieve assume role.");
+        }
+
+        $args = [
+            'region' => $region,
+            'credentials' => $sourceCreds,
+            'assume_role_params' => $this->paramsParser($this->data),
+        ];
+        $provider = new AssumeRoleCredentialProvider($args);
+        return $this->providerParams ? $args : $provider();
+    }
+
+    private function resolveSourceProfile()
+    {
+        if (!isset($this->data['source_profile'])) {
+            // No 'source_profile' information provided,
+            // Skip assumeRole, ini check in the credential chain
+            return CredentialProvider::chain(
+                CredentialProvider::ecsCredentials(),
+                CredentialProvider::instanceProfile()
+            );
+        } else {
+            $sourceProfile = $this->data['source_profile'];
+            unset($this->data['source_profile']);
+
+            $homeDir = self::getHomeDir();
+            if ($sourceProfile === $this->profile) {
+                // 'source_profile' provided is the same profile with current assumeRole profile:
+                // To avoid infinite loop, jump to default static profile (ini)
+                return CredentialProvider::chain(
+                    CredentialProvider::ini(
+                        self::DEFAULT_CREDENTIAL_PROFILE,
+                        $homeDir . self::DEFAULT_CREDENTIAL_FILE
+                    ),
+                    CredentialProvider::ini(
+                        self::DEFAULT_CONFIG_PROFILE,
+                        $homeDir . self::DEFAULT_CONFIG_FILE
+                    )
+                );
+            } else {
+                // 'source_profile' provided is a different profile:
+                // $sourceProfile could be another assumeRole profile or a static profile
+                // profile check order: assumeRole, default credential, default config
+                return CredentialProvider::chain(
+                    CredentialProvider::assumeRole($sourceProfile, $this->fileName),
+                    CredentialProvider::assumeRole(
+                        $sourceProfile,
+                        $homeDir .self::DEFAULT_CREDENTIAL_FILE
+                    ),
+                    CredentialProvider::assumeRole(
+                        $sourceProfile,
+                        $homeDir . self::DEFAULT_CONFIG_FILE
+                    ),
+                    CredentialProvider::ini($sourceProfile, $this->fileName),
+                    CredentialProvider::ini(
+                        $sourceProfile,
+                        $homeDir . self::DEFAULT_CREDENTIAL_FILE
+                    ),
+                    CredentialProvider::ini(
```

Comment:
Would all these chain calls really load the ini file and ini parse it this many times?